### PR TITLE
Put pan limits when scaling

### PR DESCRIPTION
**Pull request by tarukosu**
## Overview

Correct uv positions to prevent textures from being displayed beyond the limits.

Unity-2018 4 20f1-_PREVIEW-PACKAGES-IN-USE_-New-Scene (1)

## Changes
- Fixes: #8573

### CODE REVIEW

File: Assets/MRTK/SDK/Features/UX/Scripts/Slate/HandInteractionPanZoom.cs [hunk 1]
```diff
@@ -415,6 +415,55 @@ private void UpdateUVMapping()
                         // This is where zoom is applied if Active
                         uvs[i] = ((uvs[i] - scaleUVCentroid) / scaleUVDelta) + scaleUVCentroid;
                     }
+
+                    // Put pan limits
+                    if (!unlimitedPan)
```
Comment (vaoliva) [inline Assets/MRTK/SDK/Features/UX/Scripts/Slate/HandInteractionPanZoom.cs:420]: Would be a good idea to clamp the uv values before assigning new values.

**Review by vaoliva — COMMENTED** [review]
Hi @tarukosu ! Thanks for addressing this bug. 
Could you please also add a test covering this scenario?

**Comment by tarukosu** [comment]
@vaoliva
I've added several tests.
Could you confirm them?

**Review by vaoliva — CHANGES_REQUESTED** [review]
Trying different max min scale values while keeping max pan values the same leads to the following result:
Screenshot (72)
slate_max_pan

There is also an added jitter near the limits. 

Could you please ensure the changes also work for different max min scale and add proper tests for these scenarios?

File: Assets/MRTK/SDK/Features/UX/Scripts/Slate/HandInteractionPanZoom.cs [hunk 2]
```diff
@@ -415,6 +415,55 @@ private void UpdateUVMapping()
...
+                        }
+
+                        if (correctX || correctY)
+                        {
+                            for (int i = 0; i < uvs.Count; ++i)
+                            {
+                                uvs[i] += positionCorrection;
+                            }
+                        }
+                    }
```
Comment (vaoliva) [inline Assets/MRTK/SDK/Features/UX/Scripts/Slate/HandInteractionPanZoom.cs:466]: I have noticed there is already a check for pan limits before applying uv deltas. Could you please try avoiding the duplicates and combine the checking code and its variables.

`// Test for pan limits
            Vector2 uvDelta = new Vector2(totalUVOffset.x, -totalUVOffset.y);
            if (!unlimitedPan)
            {
                bool xLimited = false;
                bool yLimited = false;
                for (int i = 0; i < uvs.Count; ++i)
                {
                    uvTestValue = uvs[i] - uvDelta;
                    if (uvTestValue.x > tiling.x * maxPanHorizontal || uvTestValue.x < -(tiling.x * maxPanHorizontal))
                    {
                        xLimited = true;
                    }
                    if (uvTestValue.y > tiling.y * maxPanVertical || uvTestValue.y < -(tiling.y * maxPanVertical))
                    {
                        yLimited = true;
                    }
                }`

File: Assets/MRTK/Tests/PlayModeTests/SlateTests.cs [hunk 3]
```diff
@@ -275,6 +478,38 @@ private void InstantiateFromPrefab(Vector3? position = null, Quaternion? rotatio
...
+            Assert.IsNotNull(unlimitedPanField);
+            unlimitedPanField.SetValue(panZoom, false);
+
+            var lockHorizontalField = typeof(HandInteractionPanZoom).GetField("lockHorizontal", BindingFlags.NonPublic | BindingFlags.Instance);
+            Assert.IsNotNull(lockHorizontalField);
+            lockHorizontalField.SetValue(panZoom, false);
+
+            var lockVerticalField = typeof(HandInteractionPanZoom).GetField("lockVertical", BindingFlags.NonPublic | BindingFlags.Instance);
+            Assert.IsNotNull(lockVerticalField);
+            lockVerticalField.SetValue(panZoom, false);
```
Comment (vaoliva) [inline Assets/MRTK/Tests/PlayModeTests/SlateTests.cs:502]: I believe the default value for "lockvertical" and "lockhorizontal" is already false. Maybe this is not needed?

File: Assets/MRTK/Tests/PlayModeTests/SlateTests.cs [hunk 4]
```diff
@@ -222,6 +225,206 @@ public IEnumerator InstantiateGGVScroll()
...
+
+            Vector3 screenPoint = CameraCache.Main.ViewportToScreenPoint(new Vector3(0.5f, 0.0f, 0.5f));
+            for (var i = 0; i < 3; i++)
+            {
+                yield return handRight.MoveTo(CameraCache.Main.ScreenToWorldPoint(screenPoint));
+                yield return handRight.SetGesture(ArticulatedHandPose.GestureId.Pinch);
+                yield return handRight.Move(new Vector3(-0.5f, 0f, 0), 10);
+                yield return handRight.SetGesture(ArticulatedHandPose.GestureId.Open);
+            }
+
```
Comment (vaoliva) [inline Assets/MRTK/Tests/PlayModeTests/SlateTests.cs:334]: PrefabScrollRightLimited, PrefabScrollLeftLimited, PrefabScrollUpLimited and PrefabScrollDownLimited have very similar code. Could you please combine them or put duplicated code into a method?

File: Assets/MRTK/Tests/PlayModeTests/SlateTests.cs (hunk 3 above)
Comment (tarukosu) [reply Assets/MRTK/Tests/PlayModeTests/SlateTests.cs:502]: The default value depends on the settings of the slate prefab.
I prefer to explicitly specify the value in the test code.

File: Assets/MRTK/Tests/PlayModeTests/SlateTests.cs (hunk 4 above)
Comment (tarukosu) [reply Assets/MRTK/Tests/PlayModeTests/SlateTests.cs:286]: I've created ScrollToLimit method and remove duplicated code.

**Comment by tarukosu** [comment]
I refactored the scaling and scrolling code.

**Comment by RogPodge** [comment]
Did  manual pull down and check, this PR looks good to me barring our CI suite.

Sorry for being late getting back to this @tarukosu !

**Comment by RogPodge** [comment]
For some cursed reason, Unity has different UV mesh orders for 2018 vs 2019+

2018:
image

2019+:
image


Making some manual suggestions and committing those changes to fix the issue with the tests

File: Assets/MRTK/Tests/PlayModeTests/SlateTests.cs [hunk 5]
```diff
@@ -222,6 +226,178 @@ public IEnumerator InstantiateGGVScroll()
...
+        public IEnumerator PrefabScrollUpLimited()
+        {
+            var maxPan = 4f;
+            InstantiatePanLimitedSlateFromPrefab(maxPanHorizontal: maxPan, maxPanVertical: maxPan);
+
+            yield return ScrollToLimit(Vector3.up);
+
+            var uvs = new List<Vector2>();
+            meshFilter.mesh.GetUVs(0, uvs);
+            Assert.AreEqual(maxPan * material.mainTextureScale.y, uvs[1].y, 0.05, "mesh uv is not correct");
```
Comment (RogPodge) [inline Assets/MRTK/Tests/PlayModeTests/SlateTests.cs:242]:
```suggestion
#if UNITY_2019_1_OR_NEWER
            Assert.AreEqual(maxPan * material.mainTextureScale.y, uvs[3].y, 0.05, "mesh uv is not correct");
#else
            Assert.AreEqual(maxPan * material.mainTextureScale.y, uvs[1].y, 0.05, "mesh uv is not correct");
#endif
```

File: Assets/MRTK/Tests/PlayModeTests/SlateTests.cs [hunk 6]
```diff
@@ -222,6 +226,178 @@ public IEnumerator InstantiateGGVScroll()
...
+        public IEnumerator PrefabScrollRightLimited()
+        {
+            var maxPan = 4f;
+            InstantiatePanLimitedSlateFromPrefab(maxPanHorizontal: maxPan, maxPanVertical: maxPan);
+
+            yield return ScrollToLimit(Vector3.right);
+
+            var uvs = new List<Vector2>();
+            meshFilter.mesh.GetUVs(0, uvs);
+            Assert.AreEqual(maxPan * material.mainTextureScale.x, uvs[1].x, 0.05, "mesh uv is not correct");
```
Comment (RogPodge) [inline Assets/MRTK/Tests/PlayModeTests/SlateTests.cs:290]:
```suggestion
#if UNITY_2019_1_OR_NEWER
            Assert.AreEqual(maxPan * material.mainTextureScale.x, uvs[3].x, 0.05, "mesh uv is not correct");
#else
            Assert.AreEqual(maxPan * material.mainTextureScale.x, uvs[1].x, 0.05, "mesh uv is not correct");
#endif
```

File: Assets/MRTK/Tests/PlayModeTests/SlateTests.cs [hunk 7]
```diff
@@ -222,6 +226,178 @@ public IEnumerator InstantiateGGVScroll()
...
+            yield return handLeft.Show(panZoom.transform.position + Vector3.forward * -0.5f + Vector3.right * -0.2f);
+            yield return handLeft.SetGesture(ArticulatedHandPose.GestureId.Pinch);
+
+            // Use both hands to zoom out
+            yield return handLeft.Move(new Vector3(0.1f, 0.1f, 0f), 10);
+
+            var uvs = new List<Vector2>();
+            meshFilter.mesh.GetUVs(0, uvs);
+
+            Assert.AreEqual(maxPanHorizontal * material.mainTextureScale.x, uvs[1].x, 0.05, "mesh uv is not correct");
```
Comment (RogPodge) [inline Assets/MRTK/Tests/PlayModeTests/SlateTests.cs:321]:
```suggestion
#if UNITY_2019_1_OR_NEWER
            Assert.AreEqual(maxPanHorizontal * material.mainTextureScale.x, uvs[3].x, 0.05, "mesh uv is not correct");
            Assert.AreEqual(maxPanVertical * material.mainTextureScale.y, uvs[3].y, 0.05, "mesh uv is not correct");
#else
            Assert.AreEqual(maxPanHorizontal * material.mainTextureScale.x, uvs[1].x, 0.05, "mesh uv is not correct");
            Assert.AreEqual(maxPanVertical * material.mainTextureScale.y, uvs[1].y, 0.05, "mesh uv is not correct");
#endif
```

**Review by RogPodge — APPROVED** [review]

File: Assets/MRTK/Tests/PlayModeTests/SlateTests.cs [hunk 8]
```diff
@@ -222,6 +226,192 @@ public IEnumerator InstantiateGGVScroll()
...
+            meshFilter.mesh.GetUVs(0, uvs);
+
+#if UNITY_2019_1_OR_NEWER
+            Assert.AreEqual(maxPanHorizontal * material.mainTextureScale.x, uvs[3].x, 0.05, "mesh uv is not correct");
+            Assert.AreEqual(maxPanVertical * material.mainTextureScale.y, uvs[3].y, 0.05, "mesh uv is not correct");
+#else
+            Assert.AreEqual(maxPanHorizontal * material.mainTextureScale.x, uvs[1].x, 0.05, "mesh uv is not correct");
+            Assert.AreEqual(maxPanVertical * material.mainTextureScale.y, uvs[1].y, 0.05, "mesh uv is not correct");
+#endif
+            Assert.AreEqual(maxPanVertical * material.mainTextureScale.y, uvs[1].y, 0.05, "mesh uv is not correct");
```
Comment (RogPodge) [inline Assets/MRTK/Tests/PlayModeTests/SlateTests.cs:336]:
```suggestion
```

File: Assets/MRTK/SDK/Features/UX/Scripts/Slate/HandInteractionPanZoom.cs [hunk 9]
```diff
@@ -379,46 +379,69 @@ private void UpdateIdle()
...
-                    for (int i = 0; i < uvs.Count; ++i)
+                    var scaleUVDeltas = new List<Vector2>();
+                    for (int i = 0; i < uvs.Count; i++)
                     {
-                        // This is where zoom is applied if Active
-                        uvs[i] = ((uvs[i] - scaleUVCentroid) / scaleUVDelta) + scaleUVCentroid;
+                        Vector2 delta = ((uvs[i] - scaleUVCentroid) / scaleUVDelta) + scaleUVCentroid - uvs[i];
+                        scaleUVDeltas.Add(delta);
                     }
+                    UpdateUV(uvs, scaleUVDeltas);
```
Comment (keveleigh) [inline Assets/MRTK/SDK/Features/UX/Scripts/Slate/HandInteractionPanZoom.cs:405]: To double check, if we end up in this `if`, it looks like we call some sort of `UpdateUV` method twice:

1. `UpdateUV` here
2. `UpdateUVWithScroll` on line 417

Do we need to call them both?

Comment (keveleigh) [reply Assets/MRTK/SDK/Features/UX/Scripts/Slate/HandInteractionPanZoom.cs:405]: I guess if you're scaling _and_ scrolling at the same time?

File: Assets/MRTK/SDK/Features/UX/Scripts/Slate/HandInteractionPanZoom.cs [hunk 10]
```diff
@@ -429,20 +452,31 @@ private void UpdateUVMapping()
...
+                    }
+                }
+                else
+                {
+                    if (!xLimited && !yLimited)
+                    {
+                        for (int i = 0; i < uvs.Count; ++i)
+                        {
+                            uvs[i] += uvDeltas[i];
+                        }
```
Comment (keveleigh) [inline Assets/MRTK/SDK/Features/UX/Scripts/Slate/HandInteractionPanZoom.cs:469]: This case looks identical to lines 475-477. Is there any way to rework these `if`s to dedupe?

File: Assets/MRTK/SDK/Features/UX/Scripts/Slate/HandInteractionPanZoom.cs (hunk 9 above)
Comment (RogPodge) [reply Assets/MRTK/SDK/Features/UX/Scripts/Slate/HandInteractionPanZoom.cs:405]: Yeah it seems like we can do some logic juggling to make this code cleaner and avoid making multiple calls.

Going to open a new PR soon to push these larger scale changes.

File: Assets/MRTK/SDK/Features/UX/Scripts/Slate/HandInteractionPanZoom.cs (hunk 10 above)
Comment (RogPodge) [reply Assets/MRTK/SDK/Features/UX/Scripts/Slate/HandInteractionPanZoom.cs:469]: Yeah did some work but I don't think there's a good way to dedupe this without reducing the code's clarity unfortunately.

Was able to remove the duplicated calls though so see that in the followup PR #9701

**Comment by RogPodge** [comment]
Closed and resolved with #9701

**Comment by tarukosu** [comment]
@RogPodge Thank you for updating and fixing the code!

**Comment by RogPodge** [comment]
@tarukosu thank you too for opening the PR! Contributions like yours are greatly appreciated!!

**Pull request closed without merging**